libutils/proc_manager.c

```diff
@@ -0,0 +1,614 @@
+/*
+  Copyright 2021 Northern.tech AS
+
+  This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+  This program is free software; you can redistribute it and/or modify it
+  under the terms of the GNU General Public License as published by the
+  Free Software Foundation; version 3.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <platform.h>
+
+#include <map.h>                /* Map* */
+#include <alloc.h>              /* xmalloc, xasprintf */
+#include <string_lib.h>         /* StringHash_untyped, StringEqual_untyped */
+#include <logging.h>
+#include <proc_manager.h>
+
+#ifndef __MINGW32__
+#include <errno.h>
+#include <signal.h>             /* kill, SIGKILL */
+#endif
+
+#ifdef __MINGW32__
+#define fileno _fileno
+#endif
+
+#define SIGKILL_TERMINATION_TIMEOUT 5 /* seconds */
+
+struct ProcManager_ {
+    Map *procs_by_id;  /** primary storage for the subprocesses info (owns the data)    */
+    Map *procs_by_fd;  /** a different view on the data to speed up lookup by FD/stream */
+    Map *procs_by_pid; /** a different view on the data to speed up lookup by PID       */
+};
+
+static unsigned int HashFD_untyped(const void *fd, ARG_UNUSED unsigned int seed)
+{
+    assert(fd != NULL);
+    return *((unsigned int*) fd);
+}
+
+static bool FDEqual_untyped(const void *fd1, const void *fd2)
+{
+    assert((fd1 != NULL) && (fd2 != NULL));
+    return (*((int*) fd1) == (*(int*) fd2));
```

Comment:
I think it's actually clearer the way it is on the left side, saying "dereference this pointer to int" instead of "dereference this pointer type and type-cast the variable".

Reply:
So it should be changed on the right side.